### PR TITLE
Added zstd as a compression algorithm

### PR DESCRIPTION
**Pull request by V-R-Dighe**
Related https://github.com/schollz/croc/issues/504

### CODE REVIEW

**Comment by schollz** [comment]
I tested this patch against the current `flate` algorithm. This doesn't look like a clear improvement? 

Here are some tests run from `src/compress`.

Current version: 

```
> git checkout 1517767129b7f32e6e1c8d1caceace4fb8fda0be # current
> go test -bench=.
Level 9: 55% percent space savings
Level -2: 43% percent space savings
Level -2: 43% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: arm64
pkg: github.com/schollz/croc/v9/src/compress
BenchmarkCompressLevelMinusTwo-4                    1822            658837 ns/op
BenchmarkCompressLevelNine-4                        1024           1180830 ns/op
BenchmarkCompressLevelMinusTwoBinary-4                66          16841274 ns/op
BenchmarkCompressLevelNineBinary-4                    12          99928833 ns/op
```


`zstd` version:

```
> git checkout V-R-Dighe-develop
> go test -bench=.
Level 9: 53% percent space savings
Level -2: 23% percent space savings
Level -2: 23% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: arm64
pkg: github.com/schollz/croc/v9/src/compress
BenchmarkCompressLevelMinusTwo-4                   25466             47648 ns/op
BenchmarkCompressLevelNine-4                         214           4976949 ns/op
BenchmarkCompressLevelMinusTwoBinary-4                75          18479281 ns/op
BenchmarkCompressLevelNineBinary-4                     9         119355260 ns/op
```


For context, by default the level is `-2` (HuffmanOnly).

**Comment by stefins** [comment]
Hi,
I tried the benchmark with https://github.com/klauspost/compress/zstd 
**Results**
With `github.com/DataDog/zstd`
```sh
Level 9: 53% percent space savings
Level -2: 23% percent space savings
Level -2: 23% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: amd64
pkg: github.com/schollz/croc/v9/src/compress
cpu: Intel(R) Celeron(R) CPU 4205U @ 1.80GHz
BenchmarkCompressLevelMinusTwo-2         	   45229	     27757 ns/op
BenchmarkCompressLevelNine-2             	     697	   1552866 ns/op
BenchmarkCompressLevelMinusTwoBinary-2   	     408	   2936133 ns/op
BenchmarkCompressLevelNineBinary-2       	      24	  54015984 ns/op
PASS
ok  	github.com/schollz/croc/v9/src/compress	6.336s
```

With `github.com/klauspost/compress/zstd`
```sh
Level 9: 52% percent space savings
Level -2: 51% percent space savings
Level -2: 51% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: amd64
pkg: github.com/schollz/croc/v9/src/compress
cpu: Intel(R) Celeron(R) CPU 4205U @ 1.80GHz
BenchmarkCompressLevelMinusTwo-2         	    1976	    682288 ns/op
BenchmarkCompressLevelNine-2             	     151	   8247417 ns/op
BenchmarkCompressLevelMinusTwoBinary-2   	     486	   2719676 ns/op
BenchmarkCompressLevelNineBinary-2       	     154	   8582093 ns/op
PASS
ok  	github.com/schollz/croc/v9/src/compress	8.907s
```
With current flate
```sh
Level 9: 55% percent space savings
Level -2: 43% percent space savings
Level -2: 43% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: amd64
pkg: github.com/schollz/croc/v9/src/compress
cpu: Intel(R) Celeron(R) CPU 4205U @ 1.80GHz
BenchmarkCompressLevelMinusTwo-2         	    3913	    300315 ns/op
BenchmarkCompressLevelNine-2             	    2180	    523972 ns/op
BenchmarkCompressLevelMinusTwoBinary-2   	     237	   4889606 ns/op
BenchmarkCompressLevelNineBinary-2       	      24	  43074478 ns/op
PASS
ok  	github.com/schollz/croc/v9/src/compress	5.103s
```

Seems like the bottleneck is with CGO calls in https://github.com/DataDog/zstd although I'm not sure 

:)

**Comment by stefins** [comment]
Hi,
I tried the benchmark with https://github.com/klauspost/compress/tree/master/zstd
**Results**
With `github.com/DataDog/zstd`
```sh
Level 9: 53% percent space savings
Level -2: 23% percent space savings
Level -2: 23% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: amd64
pkg: github.com/schollz/croc/v9/src/compress
BenchmarkCompressLevelMinusTwo-2         	   45229	     27757 ns/op
BenchmarkCompressLevelNine-2             	     697	   1552866 ns/op
BenchmarkCompressLevelMinusTwoBinary-2   	     408	   2936133 ns/op
BenchmarkCompressLevelNineBinary-2       	      24	  54015984 ns/op
PASS
ok  	github.com/schollz/croc/v9/src/compress	6.336s
```

With `github.com/klauspost/compress/zstd`
```sh
Level 9: 52% percent space savings
Level -2: 51% percent space savings
Level -2: 51% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: amd64
pkg: github.com/schollz/croc/v9/src/compress
BenchmarkCompressLevelMinusTwo-2         	    1976	    682288 ns/op
BenchmarkCompressLevelNine-2             	     151	   8247417 ns/op
BenchmarkCompressLevelMinusTwoBinary-2   	     486	   2719676 ns/op
BenchmarkCompressLevelNineBinary-2       	     154	   8582093 ns/op
PASS
ok  	github.com/schollz/croc/v9/src/compress	8.907s
```
With current flate
```sh
Level 9: 55% percent space savings
Level -2: 43% percent space savings
Level -2: 43% percent space savings
random, Level -2: -0% percent space savings
random, Level 9: -0% percent space savings
goos: linux
goarch: amd64
pkg: github.com/schollz/croc/v9/src/compress
BenchmarkCompressLevelMinusTwo-2         	    3913	    300315 ns/op
BenchmarkCompressLevelNine-2             	    2180	    523972 ns/op
BenchmarkCompressLevelMinusTwoBinary-2   	     237	   4889606 ns/op
BenchmarkCompressLevelNineBinary-2       	      24	  43074478 ns/op
PASS
ok  	github.com/schollz/croc/v9/src/compress	5.103s
```

Seems like the bottleneck is with CGO calls in https://github.com/DataDog/zstd although I'm not sure 

:)

File: go.mod [hunk 1]
```diff
@@ -19,6 +19,7 @@ require (
 )
 
 require (
+	github.com/DataDog/zstd v1.5.2
```
Comment (stefins) [inline go.mod:22]:
```suggestion
	github.com/klauspost/compress v1.15.9
```

File: src/compress/compress.go [hunk 2]
```diff
@@ -2,9 +2,10 @@ package compress
 
 import (
 	"bytes"
-	"compress/flate"
 	"io"
 
+	"github.com/DataDog/zstd"
```
Comment (stefins) [inline src/compress/compress.go:7]:
```suggestion
	"github.com/klauspost/compress/zstd"
```

File: src/compress/compress.go [hunk 3]
```diff
@@ -30,22 +31,18 @@ func Decompress(src []byte) []byte {
...
 
-// compress uses flate to compress a byte slice to a corresponding level
+// compress uses zstd to compress a byte slice to a corresponding level
 func compress(src []byte, dest io.Writer, level int) {
-	compressor, err := flate.NewWriter(dest, level)
-	if err != nil {
-		log.Debugf("error level data: %v", err)
-		return
-	}
+	compressor := zstd.NewWriterLevel(dest, level)
```
Comment (stefins) [inline src/compress/compress.go:36]:
```suggestion
	compressor, err := zstd.NewWriter(dest, zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(level)))
	if err != nil {
		panic(err)
	}
```

File: src/compress/compress.go [hunk 4]
```diff
@@ -30,22 +31,18 @@ func Decompress(src []byte) []byte {
...
 		log.Debugf("error writing data: %v", err)
 	}
 	compressor.Close()
 }
 
-// compress uses flate to decompress an io.Reader
+// decompress uses zstd to decompress an io.Reader
 func decompress(src io.Reader, dest io.Writer) {
-	decompressor := flate.NewReader(src)
+	decompressor := zstd.NewReader(src)
```
Comment (stefins) [inline src/compress/compress.go:45]:
```suggestion
	decompressor, err := zstd.NewReader(src)
	if err != nil {
		panic(err)
	}
```

**Review by stefins — CHANGES_REQUESTED** [review]

**Comment by V-R-Dighe** [comment]
Hi @stefins,
Thanks for the suggestions.
I have reflected the changes. Please have a look. 

cc @schollz.

**Comment by schollz** [comment]
Really nice work @V-R-Dighe !! And thanks for clearing up that barrier @stefins !! This is a great improvement

**Pull request merged**